flight/PiOS/Common/pios_port.c

```diff
@@ -0,0 +1,623 @@
+                        pios_usart_dsm_id, bind)) {
+                PIOS_Assert(0);
+        }
+        
+        uintptr_t pios_dsm_rcvr_id;
+        if (PIOS_RCVR_Init(&pios_dsm_rcvr_id, &pios_dsm_rcvr_driver, pios_dsm_id)) {
+                PIOS_Assert(0);
+        }
+        pios_rcvr_group_map[MANUALCONTROLSETTINGS_CHANNELGROUPS_DSM] = pios_dsm_rcvr_id;
+}
+
+#endif
+
+#ifdef PIOS_INCLUDE_HSUM
+static void PIOS_HAL_configure_hsum(const struct pios_usart_cfg *pios_usart_hsum_cfg,
+                const struct pios_com_driver *pios_usart_com_driver,enum pios_hsum_proto *proto,
+                ManualControlSettingsChannelGroupsOptions channelgroup)
+{
+        uintptr_t pios_usart_hsum_id;
+        if (PIOS_USART_Init(&pios_usart_hsum_id, pios_usart_hsum_cfg)) {
+                PIOS_Assert(0);
+        }
+        
+        uintptr_t pios_hsum_id;
+        if (PIOS_HSUM_Init(&pios_hsum_id, pios_usart_com_driver,
+                           pios_usart_hsum_id, *proto)) {
+                PIOS_Assert(0);
+        }
+        
+        uintptr_t pios_hsum_rcvr_id;
+        if (PIOS_RCVR_Init(&pios_hsum_rcvr_id, &pios_hsum_rcvr_driver, pios_hsum_id)) {
+                PIOS_Assert(0);
+        }
+        pios_rcvr_group_map[channelgroup] = pios_hsum_rcvr_id;
+}
+#endif
+
+void PIOS_HAL_configure_port(HwSharedPortTypesOptions port_type,
+	const struct pios_usart_cfg *usart_port_cfg, 
+	const struct pios_com_driver *com_driver, 
+	uint32_t *i2c_id,
+	const struct pios_i2c_adapter_cfg *i2c_cfg,
+	const struct pios_ppm_cfg *ppm_cfg,
+	uint32_t led_id,
+/* TODO: future work to factor most of these away */
+	const struct pios_usart_cfg *usart_dsm_hsum_cfg,
+	const struct pios_dsm_cfg *dsm_cfg,
+	int dsm_bind,
+	const struct pios_usart_cfg *sbus_rcvr_cfg,
+	const struct pios_sbus_cfg *sbus_cfg,
+	bool sbus_toggle
+	)
+{
+	uintptr_t port_driver_id;
+	uintptr_t *target=NULL, *target2=NULL;;
+
+        switch (port_type) {
+		case HWSHARED_PORTTYPES_I2C:
+#if defined(PIOS_INCLUDE_I2C)
+			if (i2c_id && i2c_cfg) {
+				if (PIOS_I2C_Init(i2c_id, i2c_cfg)) {
+					PIOS_Assert(0);
+				}
+				if (PIOS_I2C_CheckClear(*i2c_id) != 0)
+					PIOS_HAL_panic(led_id, 6);
+			}
+#endif  /* PIOS_INCLUDE_I2C */
+			break;
+		case HWSHARED_PORTTYPES_PPM:
+#if defined(PIOS_INCLUDE_PPM)
+			if (ppm_cfg) { 
+				uintptr_t pios_ppm_id;
+				PIOS_PPM_Init(&pios_ppm_id, ppm_cfg);
+
+				uintptr_t pios_ppm_rcvr_id;
+				if (PIOS_RCVR_Init(&pios_ppm_rcvr_id, &pios_ppm_rcvr_driver, pios_ppm_id)) {
+					PIOS_Assert(0);
+				}
+				pios_rcvr_group_map[MANUALCONTROLSETTINGS_CHANNELGROUPS_PPM] = pios_ppm_rcvr_id;
+			}
+#endif  /* PIOS_INCLUDE_PPM */
+			break;
+
+		case HWSHARED_PORTTYPES_DISABLED:
+			break;
+		case HWSHARED_PORTTYPES_TELEMETRY:
+			PIOS_HAL_configure_com(usart_port_cfg, PIOS_COM_TELEM_RF_RX_BUF_LEN, PIOS_COM_TELEM_RF_TX_BUF_LEN, com_driver, &port_driver_id);
+			// Name of this should change XXX
+			target = &pios_com_telem_rf_id;
```

Comment:
Yah.  It's the lesser of evils since we can't really pass things to the modules.  You either need to break that, or to have a broker that you can ask "give me the id of the thing assigned to... telemetry" that modules call.  In the future the hal is going to own all of these and modules can share them.